packages/block-editor/src/store/actions.js

```diff
@@ -201,16 +220,41 @@ export function toggleSelection( isSelectionEnabled = true ) {
  *
  * @param {(string|string[])} clientIds Block client ID(s) to replace.
  * @param {(Object|Object[])} blocks    Replacement block(s).
- *
- * @return {Object} Action object.
  */
-export function replaceBlocks( clientIds, blocks ) {
-	return {
+export function* replaceBlocks( clientIds, blocks ) {
+	clientIds = castArray( clientIds );
+	blocks = castArray( blocks );
+	const rootClientId = yield select(
+		'core/block-editor',
+		'getBlockRootClientId',
+		first( clientIds )
+	);
+	// Replace is valid if the new blocks can be inserted in the root block
+	// or if we had a block of the same type in the position of the block being replaced.
```

Comment:
>or if we had a block of the same type in the position of the block being replaced.

Is there an occasion of this you had in mind that led you to implement this exception? It's not clear to me for what reason we'd want it, and it adds some complexity to the implementation.

Reply:
Hi @aduth,
Initially, I followed your suggestion, but on a second thought, I think the condition has some benefits for the merging of blocks.
In the following gist we see a sample block with InnerBlocks:
https://gist.github.com/jorgefilipecosta/6d8c1620b46d916d8a16a23e730a2ab6


If we have this condition the user can merge paragraphs by pressing backspace or delete. Without this condition, merge is not possible. But the same result is possible to achieve by removing a paragraph and appending its content to the content of another paragraph -- So I think merging should be possible.

Reply:
> In the following gist we see a sample block with InnerBlocks:
> https://gist.github.com/jorgefilipecosta/6d8c1620b46d916d8a16a23e730a2ab6

This seems like an invalid use of InnerBlocks, by the fact the template contains blocks not valid per `allowedBlocks`. At the very least, we shouldn't optimize to anticipate this use-case.

Reply:
Hi @aduth there are InnerBlocks scenarios where I expect this situation to happen.
For example, I just want to allow a maximum of 3 headings inside my InnerBlocks area. The way to implement this restriction would be remove heading from the allowed blocks if we already have 3 headings.

Reply:
But I agree even this last scenario I referred is not normal and maybe something we should not optimize for.
I thought replacements would be needed to paste a block in the same position as an existing block. Paste HTML heading in a heading block, but that's not the case. I will reflect a little bit to make sure I'm not missing anything but I find don't any "normal" case I will simplify the logic.